src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -1922,12 +1923,89 @@ static bool needsTupleElementNamesAttribute(TypeSymbol type)
                 }
 
                 var resultType = type.VisitType(
-                    predicate: (t, a, b) => !t.TupleElementNames.IsDefaultOrEmpty && !t.IsErrorType(),
+                    predicate: (t, _, _, _) => !t.TupleElementNames.IsDefaultOrEmpty && !t.IsErrorType(),
                     arg: (object?)null);
                 return resultType is object;
             }
         }
 
+        private void CheckUnboundedExtensionErasure(BindingDiagnosticBag diagnostics)
+        {
+            if (this.GetExtendedTypeNoUseSiteDiagnostics(null) is { } extendedType
+                && foundUnboundedErasure(extendedType, extensionsBeingErased: ConsList<TypeSymbol>.Empty.Prepend(this)))
+            {
+                diagnostics.Add(ErrorCode.ERR_CircularBase, Locations[0], extendedType, this);
+            }
+            return;
+
+            static bool foundUnboundedErasure(TypeSymbol type, ConsList<TypeSymbol> extensionsBeingErased, bool isContainer = false)
+            {
+                if (type is NamedTypeSymbol)
+                {
+                    if (!isContainer && type.IsExtension)
+                    {
+                        if (extensionsBeingErased.Contains(type.OriginalDefinition))
```

Comment:
>type.OriginalDefinition [](http://example.com/codeflow?start=59&length=23)

Do we have a test demonstrating that checking for specific extension constructions instead is insufficient? I.e. the one that involves infinite generic expansion, where going through erasure process we would be running into more and more unique constructions of the same extension type without an end.

Reply:
Yes, I think `TypeDepends_SelfReference_Expanding` because ever new type that result from erasure/expansion is new and larger than previous.